File: .claude.backup/skills/firebase-auth-universal/templates/mobile/ios/AuthManager.swift
/**
 * AuthManager.swift
 * Firebase Authentication Manager for iOS (Swift)
 * Complete production-ready implementation with all auth methods
 */

import Foundation
import FirebaseCore
import FirebaseAuth
import GoogleSignIn
import AuthenticationServices
import Security

// =============================================================================
// MARK: - Auth Manager
// =============================================================================

@MainActor
class AuthManager: ObservableObject {

    // MARK: - Singleton
    static let shared = AuthManager()

    // MARK: - Published Properties
    @Published var user: User?
    @Published var isAuthenticated = false
    @Published var isLoading = false
    @Published var errorMessage: String?

    // MARK: - Private Properties
    private var authStateHandle: AuthStateDidChangeListenerHandle?
    private let auth = Auth.auth()

    // MARK: - Initialization
    private init() {
        // Configure Firebase
        if FirebaseApp.app() == nil {
            FirebaseApp.configure()
        }

        // Set up auth state listener
        setupAuthStateListener()
    }

    // MARK: - Auth State Listener
    private func setupAuthStateListener() {
        authStateHandle = auth.addStateDidChangeListener { [weak self] _, user in
            guard let self = self else { return }

            Task { @MainActor in
                self.user = user
                self.isAuthenticated = user != nil

                // Save ID token to Keychain if user is authenticated
                if let user = user {
                    try? await self.saveIdTokenToKeychain()
                } else {
                    self.deleteIdTokenFromKeychain()
                }
            }
        }
    }

    deinit {
        if let handle = authStateHandle {
            auth.removeStateDidChangeListener(handle)
        }
    }

    // =============================================================================
    // MARK: - Email/Password Authentication
    // =============================================================================

    /// Sign up with email and password
    func signUp(email: String, password: String, displayName: String? = nil) async throws {
        isLoading = true
        errorMessage = nil

        defer { isLoading = false }

        do {
            let authResult = try await auth.createUser(withEmail: email, password: password)

            // Update display name if provided
            if let displayName = displayName {
                let changeRequest = authResult.user.createProfileChangeRequest()
                changeRequest.displayName = displayName
                try await changeRequest.commitChanges()
            }

            // Send email verification
            try await authResult.user.sendEmailVerification()

            self.user = authResult.user

        } catch let error as NSError {
            self.errorMessage = getErrorMessage(error)
            throw error
        }
    }

    /// Sign in with email and password
    func signIn(email: String, password: String) async throws {
        isLoading = true
        errorMessage = nil

        defer { isLoading = false }

        do {
            let authResult = try await auth.signIn(withEmail: email, password: password)
            self.user = authResult.user

        } catch let error as NSError {
            self.errorMessage = getErrorMessage(error)
            throw error
        }
    }

    /// Sign out
    func signOut() throws {
        do {
            try auth.signOut()
            self.user = nil
            self.isAuthenticated = false
            deleteIdTokenFromKeychain()

        } catch {
            throw error
        }
    }

    /// Send password reset email
    func resetPassword(email: String) async throws {
        isLoading = true
        errorMessage = nil

        defer { isLoading = false }

        do {
            try await auth.sendPasswordReset(withEmail: email)
        } catch let error as NSError {
            self.errorMessage = getErrorMessage(error)
            throw error
        }
    }

    /// Send email verification
    func sendEmailVerification() async throws {
        guard let user = auth.currentUser else {
            throw NSError(domain: "AuthManager", code: 401, userInfo: [
                NSLocalizedDescriptionKey: "No user logged in"
            ])
        }

        try await user.sendEmailVerification()
    }

    /// Change password
    func changePassword(currentPassword: String, newPassword: String) async throws {
        guard let user = auth.currentUser, let email = user.email else {
            throw NSError(domain: "AuthManager", code: 401, userInfo: [
                NSLocalizedDescriptionKey: "No user logged in"
            ])
        }

        // Re-authenticate user
        let credential = EmailAuthProvider.credential(withEmail: email, password: currentPassword)
        try await user.reauthenticate(with: credential)

        // Update password
        try await user.updatePassword(to: newPassword)
    }

    // =============================================================================
    // MARK: - Google Sign In
    // =============================================================================

    /// Sign in with Google
    func signInWithGoogle() async throws {
        guard let clientID = FirebaseApp.app()?.options.clientID else {
            throw NSError(domain: "AuthManager", code: 500, userInfo: [
                NSLocalizedDescriptionKey: "Missing Firebase client ID"
            ])
        }

        // Configure Google Sign In
        let config = GIDConfiguration(clientID: clientID)
        GIDSignIn.sharedInstance.configuration = config

        // Get root view controller
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let rootViewController = windowScene.windows.first?.rootViewController else {
            throw NSError(domain: "AuthManager", code: 500, userInfo: [
                NSLocalizedDescriptionKey: "No root view controller"
            ])
        }

        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        do {
            // Start Google Sign In flow
            let result = try await GIDSignIn.sharedInstance.signIn(withPresenting: rootViewController)

            guard let idToken = result.user.idToken?.tokenString else {
                throw NSError(domain: "AuthManager", code: 500, userInfo: [
                    NSLocalizedDescriptionKey: "Missing Google ID token"
                ])
            }

            let accessToken = result.user.accessToken.tokenString

            // Create Firebase credential
            let credential = GoogleAuthProvider.credential(
                withIDToken: idToken,
                accessToken: accessToken
            )

            // Sign in to Firebase
            let authResult = try await auth.signIn(with: credential)
            self.user = authResult.user

        } catch let error as NSError {
            self.errorMessage = getErrorMessage(error)
            throw error
        }
    }

    // =============================================================================
    // MARK: - Apple Sign In
    // =============================================================================

    /// Sign in with Apple
    func signInWithApple(nonce: String, idTokenString: String, rawNonce: String) async throws {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        do {
            // Create Firebase credential
            let credential = OAuthProvider.credential(
                withProviderID: "apple.com",
                idToken: idTokenString,
                rawNonce: rawNonce
            )

            // Sign in to Firebase
            let authResult = try await auth.signIn(with: credential)
            self.user = authResult.user

        } catch let error as NSError {
            self.errorMessage = getErrorMessage(error)
            throw error
        }
    }

    // =============================================================================
    // MARK: - Phone Authentication
    // =============================================================================

    /// Send verification code to phone number
    func sendPhoneVerificationCode(phoneNumber: String) async throws -> String {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        do {
            let verificationID = try await PhoneAuthProvider.provider()
                .verifyPhoneNumber(phoneNumber, uiDelegate: nil)
            return verificationID

        } catch let error as NSError {
            self.errorMessage = getErrorMessage(error)
            throw error
        }
    }

    /// Verify phone number with code
    func verifyPhoneNumber(verificationID: String, verificationCode: String) async throws {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        do {
            let credential = PhoneAuthProvider.provider().credential(
                withVerificationID: verificationID,
                verificationCode: verificationCode
            )

            let authResult = try await auth.signIn(with: credential)
            self.user = authResult.user

        } catch let error as NSError {
            self.errorMessage = getErrorMessage(error)
            throw error
        }
    }

    // =============================================================================
    // MARK: - Token Management
    // =============================================================================

    /// Get Firebase ID token
    func getIdToken(forceRefresh: Bool = false) async throws -> String? {
        guard let user = auth.currentUser else { return nil }

        do {
            let token = try await user.getIDToken(forcingRefresh: forceRefresh)
            return token
        } catch {
            throw error
        }
    }

    /// Refresh ID token
    func refreshToken() async throws -> String? {
        return try await getIdToken(forceRefresh: true)
    }

    // =============================================================================
    // MARK: - Keychain Management (Secure Token Storage)
    // =============================================================================

    /// Save ID token to Keychain
    private func saveIdTokenToKeychain() async throws {
        guard let token = try await getIdToken() else { return }

        let tokenData = token.data(using: .utf8)!

        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: "firebase_id_token",
            kSecValueData as String: tokenData
        ]

        // Delete existing item
        SecItemDelete(query as CFDictionary)

        // Add new item
        let status = SecItemAdd(query as CFDictionary, nil)

        if status != errSecSuccess {
            print("Failed to save token to Keychain: \(status)")
        }
    }

    /// Get ID token from Keychain
    func getIdTokenFromKeychain() -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: "firebase_id_token",
            kSecReturnData as String: true
        ]

        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)

        guard status == errSecSuccess,
              let tokenData = result as? Data,
              let token = String(data: tokenData, encoding: .utf8) else {
            return nil
        }

        return token
    }

    /// Delete ID token from Keychain
    private func deleteIdTokenFromKeychain() {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: "firebase_id_token"
        ]

        SecItemDelete(query as CFDictionary)
    }

    // =============================================================================
    // MARK: - Helper Methods
    // =============================================================================

    /// Get user-friendly error message from Firebase error
    private func getErrorMessage(_ error: NSError) -> String {
        guard let errorCode = AuthErrorCode.Code(rawValue: error.code) else {
            return "An unknown error occurred"
        }

        switch errorCode {
        case .invalidEmail:
            return "Invalid email address"
        case .emailAlreadyInUse:
            return "Email address is already in use"
        case .weakPassword:
            return "Password is too weak. Use at least 6 characters"
        case .wrongPassword:
            return "Incorrect password"
        case .userNotFound:
            return "No account found with this email"
        case .userDisabled:
            return "This account has been disabled"
        case .tooManyRequests:
            return "Too many requests. Please try again later"
        case .networkError:
            return "Network error. Please check your connection"
        case .invalidCredential:
            return "Invalid credentials. Please try again"
        default:
            return error.localizedDescription
        }
    }
}

// =============================================================================
// MARK: - User Extension
// =============================================================================

extension User {
    /// Get user's display name or email
    var displayNameOrEmail: String {
        displayName ?? email ?? "Unknown User"
    }
}

File: .claude.backup/skills/firebase-auth-universal/templates/mobile/android/AuthManager.kt
/**
 * AuthManager.kt
 * Firebase Authentication Manager for Android (Kotlin)
 * Complete production-ready implementation with all auth methods
 */

package com.example.app.auth

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKeys
import com.google.firebase.FirebaseException
import com.google.firebase.auth.*
import com.google.firebase.auth.ktx.auth
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await
import java.util.concurrent.TimeUnit

/**
 * Singleton AuthManager for Firebase Authentication
 */
class AuthManager private constructor(private val context: Context) {

    companion object {
        @Volatile
        private var instance: AuthManager? = null

        fun getInstance(context: Context): AuthManager {
            return instance ?: synchronized(this) {
                instance ?: AuthManager(context.applicationContext).also { instance = it }
            }
        }
    }

    // Firebase Auth instance
    private val auth: FirebaseAuth = Firebase.auth

    // Encrypted SharedPreferences for secure token storage
    private val encryptedPrefs: SharedPreferences by lazy {
        val masterKeyAlias = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC)

        EncryptedSharedPreferences.create(
            "firebase_auth_prefs",
            masterKeyAlias,
            context,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }

    // =============================================================================
    // MARK: - Auth State
    // =============================================================================

    /**
     * Get current user
     */
    val currentUser: FirebaseUser?
        get() = auth.currentUser

    /**
     * Check if user is authenticated
     */
    val isAuthenticated: Boolean
        get() = auth.currentUser != null

    /**
     * Auth state flow
     */
    val authStateFlow: Flow<FirebaseUser?> = callbackFlow {
        val listener = FirebaseAuth.AuthStateListener { auth ->
            trySend(auth.currentUser)

            // Save or delete token on auth state change
            auth.currentUser?.let {
                saveIdToken()
            } ?: deleteIdToken()
        }

        auth.addAuthStateListener(listener)

        awaitClose {
            auth.removeAuthStateListener(listener)
        }
    }

    // =============================================================================
    // MARK: - Email/Password Authentication
    // =============================================================================

    /**
     * Sign up with email and password
     */
    suspend fun signUp(
        email: String,
        password: String,
        displayName: String? = null
    ): Result<FirebaseUser> {
        return try {
            val result = auth.createUserWithEmailAndPassword(email, password).await()

            // Update display name if provided
            displayName?.let {
                val profileUpdates = UserProfileChangeRequest.Builder()
                    .setDisplayName(it)
                    .build()
                result.user?.updateProfile(profileUpdates)?.await()
            }

            // Send email verification
            result.user?.sendEmailVerification()?.await()

            Result.success(result.user!!)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Sign in with email and password
     */
    suspend fun signIn(email: String, password: String): Result<FirebaseUser> {
        return try {
            val result = auth.signInWithEmailAndPassword(email, password).await()
            Result.success(result.user!!)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Sign out
     */
    fun signOut() {
        auth.signOut()
        deleteIdToken()
    }

    /**
     * Send password reset email
     */
    suspend fun resetPassword(email: String): Result<Unit> {
        return try {
            auth.sendPasswordResetEmail(email).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Send email verification
     */
    suspend fun sendEmailVerification(): Result<Unit> {
        return try {
            auth.currentUser?.sendEmailVerification()?.await()
                ?: return Result.failure(Exception("No user logged in"))
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Change password
     */
    suspend fun changePassword(currentPassword: String, newPassword: String): Result<Unit> {
        return try {
            val user = auth.currentUser
                ?: return Result.failure(Exception("No user logged in"))

            val email = user.email
                ?: return Result.failure(Exception("User email not found"))

            // Re-authenticate user
            val credential = EmailAuthProvider.getCredential(email, currentPassword)
            user.reauthenticate(credential).await()

            // Update password
            user.updatePassword(newPassword).await()

            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // =============================================================================
    // MARK: - Google Sign In
    // =============================================================================

    /**
     * Sign in with Google (use with Google Sign In library)
     */
    suspend fun signInWithGoogle(idToken: String): Result<FirebaseUser> {
        return try {
            val credential = GoogleAuthProvider.getCredential(idToken, null)
            val result = auth.signInWithCredential(credential).await()
            Result.success(result.user!!)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // =============================================================================
    // MARK: - Phone Authentication
    // =============================================================================

    /**
     * Send verification code to phone number
     */
    fun sendPhoneVerificationCode(
        phoneNumber: String,
        activity: android.app.Activity,
        callbacks: PhoneAuthProvider.OnVerificationStateChangedCallbacks
    ) {
        val options = PhoneAuthOptions.newBuilder(auth)
            .setPhoneNumber(phoneNumber)
            .setTimeout(60L, TimeUnit.SECONDS)
            .setActivity(activity)
            .setCallbacks(callbacks)
            .build()

        PhoneAuthProvider.verifyPhoneNumber(options)
    }

    /**
     * Verify phone number with code
     */
    suspend fun verifyPhoneNumber(
        verificationId: String,
        verificationCode: String
    ): Result<FirebaseUser> {
        return try {
            val credential = PhoneAuthProvider.getCredential(verificationId, verificationCode)
            val result = auth.signInWithCredential(credential).await()
            Result.success(result.user!!)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // =============================================================================
    // MARK: - Social Authentication (Generic)
    // =============================================================================

    /**
     * Sign in with credential (works for any OAuth provider)
     */
    suspend fun signInWithCredential(credential: AuthCredential): Result<FirebaseUser> {
        return try {
            val result = auth.signInWithCredential(credential).await()
            Result.success(result.user!!)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // =============================================================================
    // MARK: - Token Management
    // =============================================================================

    /**
     * Get Firebase ID token
     */
    suspend fun getIdToken(forceRefresh: Boolean = false): Result<String> {
        return try {
            val user = auth.currentUser
                ?: return Result.failure(Exception("No user logged in"))

            val token = user.getIdToken(forceRefresh).await()
            val idToken = token.token
                ?: return Result.failure(Exception("Failed to get ID token"))

            Result.success(idToken)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Refresh ID token
     */
    suspend fun refreshToken(): Result<String> {
        return getIdToken(forceRefresh = true)
    }

    // =============================================================================
    // MARK: - Secure Token Storage (EncryptedSharedPreferences)
    // =============================================================================

    /**
     * Save ID token to encrypted storage
     */
    private fun saveIdToken() {
        currentUser?.getIdToken(false)?.addOnSuccessListener { result ->
            result.token?.let { token ->
                encryptedPrefs.edit()
                    .putString(KEY_ID_TOKEN, token)
                    .apply()
            }
        }
    }

    /**
     * Get ID token from encrypted storage
     */
    fun getIdTokenFromStorage(): String? {
        return encryptedPrefs.getString(KEY_ID_TOKEN, null)
    }

    /**
     * Delete ID token from encrypted storage
     */
    private fun deleteIdToken() {
        encryptedPrefs.edit()
            .remove(KEY_ID_TOKEN)
            .apply()
    }

    // =============================================================================
    // MARK: - Error Handling
    // =============================================================================

    /**
     * Get user-friendly error message from Firebase exception
     */
    fun getErrorMessage(exception: Exception): String {
        return when (exception) {
            is FirebaseAuthInvalidCredentialsException -> {
                "Invalid email or password"
            }
            is FirebaseAuthInvalidUserException -> {
                when (exception.errorCode) {
                    "ERROR_USER_NOT_FOUND" -> "No account found with this email"
                    "ERROR_USER_DISABLED" -> "This account has been disabled"
                    else -> "Invalid user"
                }
            }
            is FirebaseAuthUserCollisionException -> {
                "An account already exists with this email"
            }
            is FirebaseAuthWeakPasswordException -> {
                "Password is too weak. Use at least 6 characters"
            }
            is FirebaseAuthEmailException -> {
                "Invalid email address"
            }
            is FirebaseNetworkException -> {
                "Network error. Please check your connection"
            }
            is FirebaseTooManyRequestsException -> {
                "Too many requests. Please try again later"
            }
            else -> {
                exception.message ?: "An unknown error occurred"
            }
        }
    }

    companion object {
        private const val KEY_ID_TOKEN = "firebase_id_token"
    }
}

// =============================================================================
// MARK: - Extension Functions
// =============================================================================

/**
 * Get display name or email from FirebaseUser
 */
val FirebaseUser.displayNameOrEmail: String
    get() = displayName ?: email ?: "Unknown User"
